Modify buildConfig from web console

jhadvig opened the pull request:
The access to the edit page is through the kebab dropdown
![screenshot-4](https://cloud.githubusercontent.com/assets/1668218/12098057/f9e9c592-b31f-11e5-96a5-4365b80035b0.png)
The edit layout is in similar to the bc show page
![screenshot-5](https://cloud.githubusercontent.com/assets/1668218/12098061/fcb10b0a-b31f-11e5-9b42-5b59e8a64070.png)
After update the page is redirected to the bc show page and alert is printed.
![screenshot-6](https://cloud.githubusercontent.com/assets/1668218/12098170/af25c3c0-b320-11e5-8cd6-e2b6d1746d31.png)

jhadvig commented:
@spadgett @jwforres PTAL

smarterclayton commented:
This is not your fault, but the bar graph with two entries looks ridiculous :)

jwforres commented:
I know its probably the text we used on the create page, but i think it would be better to say "Enable the webhook build trigger"

jwforres commented:
assuming we should say from "the docker registry" and not from "Docker hub" @smarterclayton ?

jwforres commented:
i would move the "Always pull..." checkbox beneath where you specify what builder image to use

jwforres commented:
can you include screenshots for the other build types

spadgett commented on assets/app/views/browse/build-config.html:
```diff
@@ -28,6 +28,9 @@
                 ><i class="fa fa-play fa-fw" aria-hidden="true"></i> Start Build</a>
               </li>
               <li>
+                <a ng-href="{{buildConfig | editResourceURL}}" role="button"><i class="fa fa-pencil fa-fw" aria-hidden="true"></i> Edit </a>
```
Does the "Edit" menu item look strange right above "Edit (Raw)"? Not sure what to suggest :(

jhadvig replied:
To be honest I wasnt sure as well, but the dropdown was an ideal place for it

spadgett replied:
If we change the "Edit (Raw)" to "Edit YAML", the difference might be clearer. @jwforres ?

jwforres replied:
Edit YAML seems ok when you also have Edit, but when you only have Edit
YAML would you understand thats how you edit the object?

On Tue, Jan 5, 2016 at 8:42 AM, Sam Padgett <notifications@github.com>
wrote:

> In assets/app/views/browse/build-config.html
> <https://github.com/openshift/origin/pull/6528#discussion_r48845170>:
>
> > @@ -28,6 +28,9 @@
> >                  ><i class="fa fa-play fa-fw" aria-hidden="true"></i> Start Build</a>
> >                </li>
> >                <li>
> > +                <a ng-href="{{buildConfig | editResourceURL}}" role="button"><i class="fa fa-pencil fa-fw" aria-hidden="true"></i> Edit </a>
>
> If we change the "Edit (Raw)" to "Edit YAML", the difference might be
> clearer. @jwforres <https://github.com/jwforres> ?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/6528/files#r48845170>.
>

spadgett commented:
I don't mind the punctuation between dropdowns when they're in a row, but it looks odd when they're stacked. Maybe add a label above each at narrow screen widths?

![openshift_web_console](https://cloud.githubusercontent.com/assets/1167259/12101245/a691a536-b301-11e5-9e67-08571279671f.png)

spadgett commented:
Env vars should grow wider to be consistent with the create from image page. We might note that they are for both build and runtime as well (@jwforres is that accurate?).

Edit build config:

![openshift_web_console](https://cloud.githubusercontent.com/assets/1167259/12101337/1a20d0ee-b302-11e5-82cb-40217001be2d.png)

Create from image:

![openshift_web_console](https://cloud.githubusercontent.com/assets/1167259/12101367/402c0f38-b302-11e5-90a1-f77d5af0a9d3.png)

jhadvig commented:
@jwforres here are other type of builds:

- application-template-dockerbuild.json
![docker](https://cloud.githubusercontent.com/assets/1668218/12101588/d0dc9eca-b335-11e5-9cff-65244adfd29b.png)
- application-template-custombuild.json
![custom](https://cloud.githubusercontent.com/assets/1668218/12101592/d8e8742c-b335-11e5-823c-3c9d4e0e4949.png)
- application-template-pullspecbuild.json
![pullspec](https://cloud.githubusercontent.com/assets/1668218/12101603/e55cc686-b335-11e5-8ad8-98716b50a39a.png)

jhadvig commented:
@spadgett makes sense to have the label above each selectbox for narrow screen widths. thanks !

spadgett commented:
@jhadvig Regarding my comment above -- https://github.com/openshift/origin/pull/6528#issuecomment-168819878 -- I had some cached stylesheets after pulling down your change. It looks fine now. Please ignore!

spadgett commented:
I am seeing the text inside this combobox wrap at mobile screen widths, though.

![openshift_web_console](https://cloud.githubusercontent.com/assets/1167259/12116399/96d10700-b387-11e5-9fba-846055a5f695.png)

spadgett commented:
Some layout problems in the iOS emulator.

![screen shot 2016-01-05 at 8 44 40 am](https://cloud.githubusercontent.com/assets/1167259/12116547/954839fc-b388-11e5-9929-9a55e5993529.png)

spadgett commented on assets/app/scripts/controllers/create/createFromImage.js:
```diff
@@ -47,7 +47,6 @@ angular.module("openshiftConsole")
             deployOnNewImage: true,
             deployOnConfigChange: true,
-            envVars : {
             }
```
Accidental change? Syntax error here.

spadgett commented on assets/app/scripts/controllers/edit/buildConfig.js:
```diff
@@ -0,0 +1,307 @@
+            $scope.buildConfig = buildConfig;
+            $scope.updatedBuildConfig = angular.copy($scope.buildConfig);
+            $scope.buildStrategy = $filter('buildStrategy')($scope.updatedBuildConfig);
+            $scope.envVars = $filter('envVarsPair')($scope.buildStrategy.env);
```
`env` could be undefined since it is `omitempty`, which would cause runtime errors later when we pass it to the `osc-key-values` directive.

spadgett commented on assets/app/scripts/app.js:
```diff
@@ -122,6 +122,10 @@ angular
         templateUrl: 'views/browse/build-config.html',
         controller: 'BuildConfigController'
       })
+      .when('/project/:project/browse/builds/:buildconfig/edit', {
```
I like the URL, but it would conflict with the route below if we ever had a build named `edit`.

jhadvig replied:
how about `/project/:project/browse/builds/edit/:buildconfig` ?? I know its not 100% REST friendly but could serve for this purpose.

spadgett replied:
Conflicts with a build config named `edit` :(

Maybe `/project/edit/builds/:buildconfig`?

spadgett replied:
That way we have the verbs `browse` and `edit` in the same spot in the URL.

jhadvig replied:
:+1:

spadgett commented on assets/app/scripts/controllers/edit/buildConfig.js:
```diff
@@ -0,0 +1,307 @@
+            
+
+            var buildFrom = $scope.buildStrategy.from;
+            var testFrom = $filter('imageObjectRef')($filter('buildStrategy')($scope.buildConfig).from, buildConfig.metadata.namespace)
```
Multiple calls to `$filter` on one line are hard to read. I'd probably just assign the filters to variables above.

jhadvig commented:
@spadgett I've updated the style for the narrow displays as suggested. Please check the screens. 
![1](https://cloud.githubusercontent.com/assets/1668218/12117140/e025466e-b3be-11e5-9a5c-427f3020f6fd.png)
![2](https://cloud.githubusercontent.com/assets/1668218/12117143/e195c4a6-b3be-11e5-91d2-d7979f297da9.png)

spadgett commented:
@jhadvig your last screenshots look good, thanks

jwforres commented:
"Execute the docker build with the "--no-cache=true" flag"  as an end user what does this mean to me, i know it literally means it will run docker build with that flag, but can we explain that a bit better as to how that affects the docker build

spadgett commented on assets/app/scripts/controllers/edit/buildConfig.js:
```diff
@@ -0,0 +1,307 @@
+              $scope.options.noCache = false;
+              if ($scope.buildConfig.spec.strategy.dockerStrategy.noCache === true) {
+                $scope.options.noCache = true;
+              };
```
Maybe

`$scope.options.noCache = $scope.buildConfig.spec.strategy.dockerStrategy.noCache;`

spadgett commented on assets/app/scripts/controllers/edit/buildConfig.js:
```diff
@@ -0,0 +1,307 @@
+
+            $scope.options.forcePull = false;
+            if ($scope.buildStrategy.forcePull == true) {
+              $scope.options.forcePull = true;
```
```
$scope.options.forcePull = $scope.buildStrategy.forcePull;
```

or if you need to make sure it's a boolean

```
$scope.options.forcePull = !!$scope.buildStrategy.forcePull;
```

This could be set in the `$scope.options` declaration above.

jhadvig replied:
great ! thahts just what I needed, thanks ! :+1:

jwforres commented:
also when you are testing make sure you edit build configs that don't actually use git source, that use one of the other source types and make sure nothing gets messed up with the source when you edit something else (like the from image) and save the form

spadgett commented on assets/app/scripts/controllers/edit/buildConfig.js:
```diff
@@ -0,0 +1,307 @@
+          kind: $scope.options.pickedBuildFromType,
+          namespace: $scope.options.pickedBuildFromNamespace,
+          name: $scope.options.pickedBuildFromImageStream + ":" + $scope.options.pickedBuildFromImageStreamTag
+        }
```
;

spadgett commented on assets/app/scripts/controllers/edit/buildConfig.js:
```diff
@@ -0,0 +1,307 @@
+        from = {
+          kind: $scope.options.pickedBuildFromType,
+          name: $scope.options.pickedBuildFromDockerImage
+        }
```
;

spadgett commented on assets/app/scripts/controllers/edit/buildConfig.js:
```diff
@@ -0,0 +1,307 @@
+        kind: "ImageStreamTag",
+        namespace: $scope.options.pickedPushToNamespace,
+        name: $scope.options.pickedPushToImageStream + ":" + $scope.options.pickedPushToImageStreamTag
+      }
```
;

spadgett commented on assets/app/scripts/controllers/edit/buildConfig.js:
```diff
@@ -0,0 +1,307 @@
+        }
+
+        var projectImageStreams = imageStreams.by("metadata.name");
+        angular.forEach(projectImageStreams, function(value, key) {
```
Maybe

```
angular.forEach(projectImageStreams, function(imageStream, name) {
```

`value` and `key` are pretty generic names.

jhadvig replied:
+1

spadgett commented on assets/app/scripts/controllers/edit/buildConfig.js:
```diff
@@ -0,0 +1,307 @@
+        var projectImageStreams = imageStreams.by("metadata.name");
+        angular.forEach(projectImageStreams, function(value, key) {
+          var tagList = [];
+          if (angular.isDefined(value.status.tags)) {
```
Any reason to not just say

```
if (value.status.tags) { ... }
```

?

jhadvig replied:
none

spadgett replied:
Or you could just call `angular.forEach`, which handles undefined for you.

```
            angular.forEach(value.status.tags, function(item){ 
              tagList.push(item["tag"]); 
            });
```

spadgett commented on assets/app/scripts/controllers/edit/buildConfig.js:
```diff
@@ -0,0 +1,307 @@
+    // updateImageStreams creates/updates the list of imageStreams and ImageStreamTags from picked namespace. As parameter takes
+    // the picked namespace, buildFrom as a boolean that indicates whether the lists are for the builder image 'picker' or the 
+    // output imageStream 'picker'.
+    $scope.updateImageStreams = function(projectName, buildFrom, firstRun) {
```
I suggest having `updateBuilderImageStreams` and `updateOutputImageStreams` scope functions that call a common helper. Then it's more obvious in the calling code what it's doing, and one less parameter each time you call them.

spadgett replied:
I'd also change `firstRun` to something like `clearSelectedOption` that makes it more clear what it's doing.

jhadvig replied:
I was thinking of splitting the `updateImageStreams` into two separate function but that would cause code dups. but I guess its ok in this case since it would make it more clearer. Thanks !

spadgett replied:
You could keep this function and make it a helper instead of a scope function. Then the two clearer named functions call the helper?

spadgett commented on assets/app/scripts/controllers/edit/buildConfig.js:
```diff
@@ -0,0 +1,307 @@
+
+    }
+
+    $scope.updateTags = function(imageStream, buildFrom) {
```
Like above, I think this is better as two functions instead of taking a `buildFrom` param.

spadgett replied:
Maybe call it `clearSelectedTag` as well?

jhadvig replied:
+1 for the `clearSelectedTag`

spadgett commented on assets/app/scripts/controllers/edit/buildConfig.js:
```diff
@@ -0,0 +1,307 @@
+
+    $scope.save = function() {
+      // Update Configuration
+      $filter('buildStrategy')($scope.updatedBuildConfig).forcePull = $scope.options.forcePull;
```
I'd assign `buildStrategy` to a variable so you don't have to call the filter so many times.

spadgett replied:
```
var buildStrategy = $filter('buildStrategy')($scope.updatedBuildConfig);
```

spadgett commented on assets/app/scripts/controllers/edit/buildConfig.js:
```diff
@@ -0,0 +1,307 @@
+              type: "GitHub"
+            }
+          );
+        }
```
Maybe... (untested)

```
var webhook = _.find($scope.buildConfig.spec.triggers, { type: "GitHub" });
if (!webhook) {
  webhook =  {
    github: {
      secret: ApplicationGenerator._generateSecret()
    },
    type: "GitHub"
  };
}
triggers.push(webhook);
```

https://lodash.com/docs#find

jhadvig replied:
hmm looks like a game-changer :ok_hand:

spadgett commented on assets/app/scripts/controllers/edit/buildConfig.js:
```diff
@@ -0,0 +1,307 @@
+        if (trigger.type === "Generic") {
+          triggers.push(trigger);
+        }
+      });
```
I think this should work (untested).

```
triggers.push(_.filter($scope.buildConfig.spec.triggers, { type: "Generic" });
```

spadgett commented on assets/app/scripts/controllers/edit/buildConfig.js:
```diff
@@ -0,0 +1,307 @@
+
+      // Update triggers
+      var triggers = [];
+      if ($scope.triggers["webhook"]) {
```
You can test `$scope.triggers.webhook` without using the bracket notation here and below.

spadgett commented on assets/app/scripts/controllers/edit/buildConfig.js:
```diff
@@ -0,0 +1,307 @@
+      });
+      if ($scope.triggers["imageChange"]) {
+        triggers.push({
+          imageChange: {},
```
Are we potentially losing data here? It looks like there is a `lastTriggeredImageID` property.

spadgett replied:
If a previous image change trigger was set.

spadgett commented on assets/app/scripts/directives/editFormLink.js:
```diff
@@ -0,0 +1,53 @@
+'use strict';
```
Looks like this directive is no longer used?

spadgett commented on assets/app/scripts/filters/resources.js:
```diff
@@ -242,6 +242,15 @@ angular.module('openshiftConsole')
+      var envVars = {};
+      angular.forEach(configEnvVars, function(value, key) {
+        envVars[configEnvVars[key].name] = configEnvVars[key].value
+      });
```
Does this work?

```
      angular.forEach(configEnvVars, function(var) {
        envVars[var.name] = var.value
      });
```

jhadvig replied:
Why shouldn't it ?

spadgett replied:
Never sure when I type code into a GitHub comment :)

jhadvig replied:
:-D

spadgett commented on assets/app/styles/_core.less:
```diff
@@ -832,6 +832,60 @@ a.disabled-link {
+      display: inline-block;
+    }
+    .multiselector-label {
+      display: none;
```
You can use the Bootstrap responsive classes to show and hide the label.

http://getbootstrap.com/css/#responsive-utilities-classes

spadgett replied:
Same for `punctuation` below.

jhadvig replied:
I've used those classes before, but for some reason (to me unknown)  I forgot about it. Thanks :+1:

spadgett commented on assets/app/views/browse/edit-build-config.html:
```diff
@@ -0,0 +1,257 @@
+
+      <div>
+        <h1>
+          Edit {{buildConfig.metadata.name}} Build Config
```
Edit Build Config {{buildConfig.metadata.name}} ?

jhadvig replied:
:+1:

spadgett commented on assets/app/views/browse/edit-build-config.html:
```diff
@@ -0,0 +1,257 @@
+                            </div>
+                          </div>
+                          <div>
+                            <span class="text-warning" ng-if="form.sourceUrl.$dirty && !sourceURLPattern.test(updatedBuildConfig.sourceUrl)">Git repository should be a URL.</span>
```
Where is `sourceURLPattern` defined?

spadgett commented on assets/app/views/browse/edit-build-config.html:
```diff
@@ -0,0 +1,257 @@
+                                name="sourceUrl"
+                                ng-model="updatedBuildConfig.spec.source.git.uri"
+                                type="text"
+                                aria-describedby="from_source_help">
```
Remove `aria-describedby` attribute since there's no help here.

spadgett commented on assets/app/views/browse/edit-build-config.html:
```diff
@@ -0,0 +1,257 @@
+                                name="sourceRef"
+                                type="text"
+                                ng-model="updatedBuildConfig.spec.source.git.ref"
+                                aria-describedby="from_source_help"
```
Remove `aria-describedby`.

spadgett replied:
Remove `aria-describedby`.

spadgett commented on assets/app/views/browse/edit-build-config.html:
```diff
@@ -0,0 +1,257 @@
+                                type="text"
+                                ng-model="updatedBuildConfig.spec.source.git.ref"
+                                aria-describedby="from_source_help"
+                                placeholder="{{updatedBuildConfig.spec.source.git.ref === undefined ? 'master' : updatedBuildConfig.spec.source.git.ref}}">
```
We shouldn't set the placeholder text to an actual value. Let's always use "master".

jhadvig replied:
right if no ref is set show the placeholder with the `master` otherwise type the ref. Makes sense

spadgett commented on assets/app/views/browse/edit-build-config.html:
```diff
@@ -0,0 +1,257 @@
+                          <div ng-if="updatedBuildConfig.spec.source.type == 'Git'">
+                            <label for="sourceRef">Git Repository Ref</label>
+                            <div>
+                              <!-- Unfortunately, we can't set type="url" because some valid values don't pass browser validation. -->
```
Remove comment since ref isn't a URL.

spadgett commented on assets/app/views/browse/edit-build-config.html:
```diff
@@ -0,0 +1,257 @@
+                          <div ng-if="updatedBuildConfig.spec.source.type == 'Git'">
+                            <label for="sourceContextDir">Git Context Dir</label>
+                            <div>
+                              <!-- Unfortunately, we can't set type="url" because some valid values don't pass browser validation. -->
```
Remove comment.

spadgett commented on assets/app/views/browse/edit-build-config.html:
```diff
@@ -0,0 +1,257 @@
+                                name="sourceContextDir"
+                                type="text"
+                                ng-model="updatedBuildConfig.spec.source.git.contextDir"
+                                aria-describedby="from_source_help"
```
Remove `aria-describedby`.

spadgett commented on assets/app/views/browse/edit-build-config.html:
```diff
@@ -0,0 +1,257 @@
+                        </div>
+
+                        <div class="form-group">
+                          <div class="checkbox" ng-if="buildConfig.spec.strategy.type === 'Docker'">
```
Nit: I'd put the `ng-if` on the parent `form-group` div so an empty div isn't added to the page for other strategies.

jhadvig replied:
+1

spadgett commented on assets/app/views/browse/edit-build-config.html:
```diff
@@ -0,0 +1,257 @@
+                              <a href data-toggle="tooltip" data-placement="right" data-original-title="This image can build.">
+                              <i class="pficon pficon-help"></i>
+                              </a>
+                            </span> -->
```
Remove commented out code?

spadgett commented on assets/app/views/browse/edit-build-config.html:
```diff
@@ -0,0 +1,257 @@
+
+                    <div class="section">
+                      <h3>Environment Variables<span class="help action-inline">
+                        <a href data-toggle="tooltip" data-placement="right"
```
I realize you probably got this from `fromimage.html`, but I'm not sure we should wrap the icon in an anchor because a screen reader would think it's a link. Do the `data-*` attributes work directly on the `i` element?

jhadvig replied:
yep they do, if we leave the the icon wrapped in the anchor and user clicks on the anchor the tooltip will still be visible, but if the icon is place outside the anchor, the tooltip will be hidden the moment the curcor leaves the anchor, which I guess is the desired case. right ?
If so I can update the logic on all other places :)

spadgett replied:
I'm not sure of best practices here for screen readers. I think it's OK. We might need to revisit tooltips anyway for #5988.

spadgett commented on assets/app/views/browse/edit-build-config.html:
```diff
@@ -0,0 +1,257 @@
+                      <h3>Environment Variables<span class="help action-inline">
+                        <a href data-toggle="tooltip" data-placement="right"
+                           data-original-title="Environment variables are used to configure and pass information to running containers.  These environment variables will be available during your build and at runtime.">
+                        <i class="pficon pficon-help"></i>
```
`aria-hidden="true"`

spadgett replied:
Also add to the other icons below.

spadgett commented on assets/app/views/browse/edit-build-config.html:
```diff
@@ -0,0 +1,257 @@
+                  <!-- unable to use form.valid.  need to fix validators in labels and key values directive -->
+                  <a class="btn btn-primary btn-lg"
+                      ng-click="save()"
+                      ng-disabled="form.invalid"
```
Does `ng-disabled` work on a link? It might be better to make this a button anyway so that enter submits the form.

spadgett commented:
Add the following attributes to text inputs so that they work well on iOS.

```
autocorrect="off"
autocapitalize="off"
spellcheck="false"
```

spadgett commented:
I'd add a fieldset around the form to disable it after clicking save so you can't double click the save button. See #5436.

spadgett commented:
Need to make sure we have `ng-required` on all required fields and that the save button is disabled if anything is missing.

spadgett commented on assets/app/views/browse/edit-build-config.html:
```diff
@@ -0,0 +1,257 @@
+                            class="form-control simpleselector"
+                            ng-model="options.pickedBuildFromType"
+                            ng-options="type.id as type.title for type in buildFromTypes">
+                          </select>
```
This is the only form field to the right of the label. Maybe put it below?

jhadvig replied:
It is already below.. I've done that in the second commit

spadgett replied:
Thanks, looks good. (Nit: We should remove the `:` from the label.)

spadgett commented on assets/app/scripts/controllers/edit/buildConfig.js:
```diff
@@ -0,0 +1,307 @@
+      DataService.list("imagestreams", {namespace: projectName}, function(imageStreams) {
+        if (buildFrom) {
+          if (!firstRun) {
+            $scope.options.pickedBuildFromImageStream = "";
```
Should we just pick the first value rather than `""` which is never valid? (Same comment for tags.)

jhadvig replied:
+1

spadgett commented on assets/app/views/browse/edit-build-config.html:
```diff
@@ -0,0 +1,257 @@
+                                ng-model="options.pickedBuildFromImageStream"
+                                ng-options="name as name for name in buildFrom.imageStreams track by name"
+                                ng-change="updateTags(options.pickedBuildFromImageStream, true)">
+                              </select>
```
I think we need an error message below saying there are no image streams for the namespace when the list is empty.

jwforres replied:
You should be able to type in one that doesn't exist (I'm assuming that was
implemented, havent tried it), so don't make it an error, make it a warning.

On Tue, Jan 5, 2016 at 1:04 PM, Sam Padgett <notifications@github.com>
wrote:

> In assets/app/views/browse/edit-build-config.html
> <https://github.com/openshift/origin/pull/6528#discussion_r48874909>:
>
> > +                              <select
> > +                                class="form-control multiselector select-builder-namespace"
> > +                                ng-model="options.pickedBuildFromNamespace"
> > +                                ng-options="name for name in buildFrom.projects track by name"
> > +                                ng-change="updateImageStreams(options.pickedBuildFromNamespace, true, false)">
> > +                              </select>
> > +                            </div>
> > +                            <span class="punctuation">/</span>
> > +                            <div class="label-with-selector">
> > +                              <label class="multiselector-label">Image Stream</label>
> > +                              <select
> > +                                class="form-control multiselector select-builder-name"
> > +                                ng-model="options.pickedBuildFromImageStream"
> > +                                ng-options="name as name for name in buildFrom.imageStreams track by name"
> > +                                ng-change="updateTags(options.pickedBuildFromImageStream, true)">
> > +                              </select>
>
> I think we need an error message below saying there are no image streams
> for the namespace when the list is empty.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/6528/files#r48874909>.
>

spadgett replied:
> You should be able to type in one that doesn't exist

What did we decide about that during the design review? I can't remember.

jwforres replied:
I thought agreed you could type it in, somehow, whether its in those inputs
or a separate input.  We can't break if someone creates a build config from
the CLI that points to an IS tag that doesn't yet exist, and then goes and
edits that BC in the web console.

On Tue, Jan 5, 2016 at 1:16 PM, Sam Padgett <notifications@github.com>
wrote:

> In assets/app/views/browse/edit-build-config.html
> <https://github.com/openshift/origin/pull/6528#discussion_r48876185>:
>
> > +                              <select
> > +                                class="form-control multiselector select-builder-namespace"
> > +                                ng-model="options.pickedBuildFromNamespace"
> > +                                ng-options="name for name in buildFrom.projects track by name"
> > +                                ng-change="updateImageStreams(options.pickedBuildFromNamespace, true, false)">
> > +                              </select>
> > +                            </div>
> > +                            <span class="punctuation">/</span>
> > +                            <div class="label-with-selector">
> > +                              <label class="multiselector-label">Image Stream</label>
> > +                              <select
> > +                                class="form-control multiselector select-builder-name"
> > +                                ng-model="options.pickedBuildFromImageStream"
> > +                                ng-options="name as name for name in buildFrom.imageStreams track by name"
> > +                                ng-change="updateTags(options.pickedBuildFromImageStream, true)">
> > +                              </select>
>
> You should be able to type in one that doesn't exist
>
> What did we decide about that during the design review? I can't remember.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/6528/files#r48876185>.
>

spadgett commented:
@jhadvig A lot of comments here, but the biggest items are

1. Layout issues on mobile (which I think you've fixed).
1. Form validation.
1. Handling a build config that points to an image stream tag that doesn't exist. We can't lose data.

spadgett commented on assets/app/scripts/controllers/edit/buildConfig.js:
```diff
@@ -0,0 +1,312 @@
+            };
+
+            if ($scope.buildConfig.spec.strategy.type === "Docker") {
+              $scope.options.noCache = !!$scope.buildConfig.spec.strategy.dockerStrategy.noCache
```
;

spadgett commented on assets/app/scripts/filters/resources.js:
```diff
@@ -242,6 +242,15 @@ angular.module('openshiftConsole')
+    return function(configEnvVars) {
+      var envVars = {};
+      angular.forEach(configEnvVars, function(env) {
+        envVars[env.name] = env.value
```
;

spadgett commented:
Looks like you're using a different breakpoint for switching to labels vs. switching to a stacked display. Feel free to roll back the change to use Bootstrap `hidden-*` and `visible-*` classes. Better to have one breakpoint in one file. (My suggestion was bad, sorry.)

![openshift_web_console](https://cloud.githubusercontent.com/assets/1167259/12154963/991e6124-b490-11e5-9359-b79bdbc0b730.png)

spadgett commented on assets/app/views/browse/edit-build-config.html:
```diff
@@ -0,0 +1,268 @@
+                                id="sourceContextDir"
+                                name="sourceContextDir"
+                                type="text"
+                                ng-model="updatedBuildConfig.spec.source.git.contextDir"
```
`contextDir` is directly under `spec.source`, not under `git`.

```
ng-model="updatedBuildConfig.spec.source.contextDir"
```

spadgett commented on assets/app/scripts/controllers/edit/buildConfig.js:
```diff
@@ -0,0 +1,312 @@
+        }
+        triggers.push(webhook);
+      }
+      triggers.push(_.find($scope.buildConfig.spec.triggers, function(obj) { return obj.type == "Generic" }));
```
We might change these calls to `_.find` to use `_.filter` in case there is more than one trigger with the same type. (Sorry if that's not what I suggested before.)

```
triggers.concat(_.filter($scope.buildConfig.spec.triggers, function(obj) { return obj.type == "Generic" }));
```

jhadvig replied:
Makes sense :)

spadgett commented on assets/app/scripts/controllers/edit/buildConfig.js:
```diff
@@ -0,0 +1,312 @@
+            imageChange: {},
+            type: "ImageChange"
+          });
+        }
```
Existing image change triggers aren't going to be added back to the triggers array, so we could lose data on edit here.

How about

```
    var imageChange = _.find($scope.buildConfig.spec.triggers, function(obj) { return obj.type == "ImageChange" });
    if (!imageChange) {
      imageChange = {
        imageChange: {},
        type: "ImageChange"
      };
    }
    triggers.push(imageChange);
 ```
 
 or if you use `_.filter`
 
```
    var imageChangeTriggers = _.filter($scope.buildConfig.spec.triggers, function(obj) { return obj.type == "ImageChange" });
    if (!imageChangeTriggers.length) {
      imageChangeTriggers.push({
        imageChange: {},
        type: "ImageChange"
      });
    }
    triggers.concat(imageChangeTriggers);
 ```
 
 (untested)

spadgett commented on assets/app/scripts/controllers/edit/buildConfig.js:
```diff
@@ -0,0 +1,336 @@
+    $scope.buildConfig = null;
+    $scope.alerts = {};
+    $scope.emptyMessage = "Loading...";
+    $scope.sourceURLPattern = /^((ftp|http|https|git):\/\/(\w+:{0,1}[^\s@]*@)|git@)?([^\s@]+)(:[0-9]+)?(\/|\/([\w#!:.?+=&%@!\-\/]))?$/;
```
We should avoid having this regex in two places, here and in createFromImage.js. It could go in ApplicationGenerator service maybe?

spadgett commented on assets/app/views/browse/edit-build-config.html:
```diff
@@ -0,0 +1,266 @@
+                            </div>
+                          </div>
+                          <div>
+                            <span class="text-warning" ng-if="form.sourceUrl.$dirty && !sourceURLPattern.test(updatedBuildConfig.sourceUrl)">Git repository should be a URL.</span>
```
```
!sourceURLPattern.test(updatedBuildConfig.spec.source.git.uri)
```

jhadvig replied:
:+1:

spadgett commented:
Around ~1200px wide, the dropdowns can wrap onto the next line.

![openshift_web_console](https://cloud.githubusercontent.com/assets/1167259/12177984/04e41e42-b53d-11e5-8538-678686eab56b.png)

spadgett commented:
I'm seeing a runtime error that I'm having trouble tracking down. It seems to come from the `valuesNotIn` filter used by `osc-key-values`.

```
TypeError: Cannot read property 'split' of undefined
    at util.js:207
    at $parseFilter (angular.js:12142)
    at Object.regularInterceptedExpression (angular.js:12832)
    at Scope.$digest (angular.js:14217)
    at Scope.$apply (angular.js:14488)
    at done (angular.js:9646)
    at completeRequest (angular.js:9836)
    at XMLHttpRequest.requestLoaded (angular.js:9777)
```

spadgett commented:
Looks like if you change

``` 
<div class="tile" ng-if="buildConfig">
```

to

```
<div class="tile" ng-show="buildConfig">
```

the runtime error goes away. I think there are bugs in how the `osc-key-values` directive is handling scope unrelated to your change.

jhadvig commented:
@spadgett both issues are taken care of. Thanks.
Also there I've found out that there is an issue (probably with ui-bootstrap version) that causes the dropdown for the secondary action to be empty, without any actions https://github.com/openshift/origin/issues/6579

jhadvig commented:
@spadgett @jwforres I've updated the PR. 
Now all tource types are editable + builder & output image can be defined as desired. If user won't find desired imageStreamTag(for builder/output/imageSource) he can choose the '-- custom --' option from the selectbox in which case the selectbox will be replaced by input field in which user can type any tag he wanted, or switch back to the selectox by pressing close icon.
For Dockerfile editation I've used ace editor as Sam proposed. Thanks ! :+1: 
Attaching the screens:
1. ImageSource editation
![screenshot-7](https://cloud.githubusercontent.com/assets/1668218/12355178/4a023eaa-bb9b-11e5-82d0-c0a6a3e4cea4.png)
2. Dockerfile editation
![screenshot-8](https://cloud.githubusercontent.com/assets/1668218/12355190/5dd8fdba-bb9b-11e5-86a4-7e25ee45f303.png)
3. Picking custom Tag
![screenshot-9](https://cloud.githubusercontent.com/assets/1668218/12355197/690bf5b6-bb9b-11e5-8598-513b86a0777e.png)
4. Entering custom Tag
![screenshot-10](https://cloud.githubusercontent.com/assets/1668218/12355208/76c0a7c4-bb9b-11e5-9d72-8c0caaf8c1e0.png)

spadgett commented on assets/app/scripts/controllers/edit/buildConfig.js:
```diff
@@ -0,0 +1,617 @@
+              $scope.options.pickedPushToDockerImage = di;
+            }
+
+            if (!!$scope.buildStrategy.from) {
```
```
if ($scope.buildStrategy.from) {
```

should work the same.

spadgett commented on assets/app/scripts/controllers/edit/buildConfig.js:
```diff
@@ -0,0 +1,617 @@
+              setBuildFromVariables("None", buildConfig.metadata.namespace, "", "", "");
+            }
+
+            if (!!$scope.updatedBuildConfig.spec.output.to) {
```
```
if ($scope.updatedBuildConfig.spec.output.to) {
```

spadgett commented on assets/app/scripts/controllers/edit/buildConfig.js:
```diff
@@ -0,0 +1,617 @@
+              // in the picker, so when the user changes the reference kind to ImageStreamTag the picker is filled with
+              // default(first) value.
+              var builderSelectFirstOption = ($scope.options.pickedBuildFromType === "DockerImage") ? true : false;
+              var outputSelectFirstOption = ($scope.options.pickedPushToType === "DockerImage") ? true : false;
```
```
var builderSelectFirstOption = $scope.options.pickedBuildFromType === "DockerImage";
var outputSelectFirstOption = $scope.options.pickedPushToType === "DockerImage";
```

jhadvig replied:
:+1:

spadgett commented on assets/app/scripts/controllers/edit/buildConfig.js:
```diff
@@ -0,0 +1,617 @@
+
+              if ($scope.sources.image) {
+                $scope.imageSourceBuildFrom.projects = angular.copy($scope.buildFrom.projects);
+                var imageSourceSelectFirstOption = ($scope.options.pickedImageSourceType === "DockerImage") ? true : false;
```
```
var imageSourceSelectFirstOption = $scope.options.pickedImageSourceType === "DockerImage";
```

spadgett commented on assets/app/scripts/controllers/edit/buildConfig.js:
```diff
@@ -0,0 +1,617 @@
+
+              angular.element(document).ready(function () {
+                // $('.combobox').combobox();
+              });
```
Let's remove this since we're aren't using it.

spadgett commented on assets/app/scripts/controllers/edit/buildConfig.js:
```diff
@@ -0,0 +1,617 @@
+                $scope.alerts["deleted"] = {
+                  type: "warning",
+                  message: "This build configuration has been deleted."
+                };
```
Should we disable the form when the config is deleted from under us?

spadgett replied:
Actually I'm not sure we should be watching at all. What happens when the build config changes as I'm editing it?

jhadvig replied:
Then it won't let you edit it at all, and redirects you to the bc browse page with a error notification

jhadvig replied:
But yeah, we should disable the form if deleted

spadgett replied:
> Then it won't let you edit it at all, and redirects you to the bc browse page with a error notification

It redirects you immediately? That might be a bit jarring. Also, if I might have made some edits I'd want a chance to copy somewhere before losing them (for instance, to the Dockerfile).

spadgett commented on assets/app/scripts/controllers/edit/buildConfig.js:
```diff
@@ -0,0 +1,617 @@
+        }, 30, function() {
+          editor.resize();
+        });
+      };
```
The function is `updateEditorWidth`, but we seem to be changing the height. Since there are many other fields on the page, we can probably use a fixed height and remove the resize handler altogether.

jhadvig replied:
yeah the resize handler is for the editor width. Mistyped the func name

spadgett replied:
> yeah the resize handler is for the editor width. Mistyped the func name

It looks like we're settings the height though in the call to `animate`. Does the width resizing work without the handling, just using CSS width?

jhadvig replied:
yes it does. will remove the function since its worthless. Thanks

spadgett commented on assets/app/scripts/controllers/edit/buildConfig.js:
```diff
@@ -0,0 +1,617 @@
+              details: $filter('getErrorDetails')(result)
+            }
+          });
+          $location.path(Navigate.resourceURL($scope.updatedBuildConfig, "BuildConfig", $scope.updatedBuildConfig.metadata.namespace, "browse"));
```
I think it would be better to stay on the page to give the user a chance to fix the problem (assuming something is wrong with the data) rather than navigating away and losing any edits.

spadgett commented on assets/app/scripts/directives/oscKeyValues.js:
```diff
@@ -22,6 +22,9 @@ angular.module("openshiftConsole")
     $scope.allowDelete = function(value){
+      if ($scope.canBeEmpty && (Object.keys($scope.entries).length === 1)) {
+        return false;
+      }
```
Maybe I have this backwards, but I would expect delete to be allowed if `canBeEmpty` is true.

jhadvig replied:
yeap, I've had it the opposite way but re-wrote it. Thanks for noticing.

spadgett commented on assets/app/scripts/directives/oscKeyValues.js:
```diff
@@ -139,7 +144,8 @@ angular.module("openshiftConsole")
         keyValidationTooltip: "@",
-        valueValidationTooltip: "@"
+        valueValidationTooltip: "@",
+        canBeEmpty: "@",
```
If you use `canBeEmpty: "="`, you shouldn't need the special boolean handling for `canBeEmpty` in `compile` below.

spadgett replied:
Or `canBeEmpty: "=?"` if optional.

spadgett commented on assets/app/scripts/directives/oscKeyValues.js:
```diff
@@ -148,6 +154,13 @@ angular.module("openshiftConsole")
+          attrs.canBeEmpty = true;
+        }else{
+          attrs.canBeEmpty = false;
+        }
```
Should the default be `true` if not set?

jhadvig replied:
it should

spadgett replied:
I'd probably flip it then to what you must have had: `preventEmpty`, then the default if not set is more obvious.

spadgett commented on assets/app/scripts/services/navigate.js:
```diff
@@ -54,7 +54,7 @@ angular.module("openshiftConsole")
       // name and buildConfig (deploymentConfig) name, then we will need either a specialized method for that, or an
       // additional opts param for extra opts.
-      resourceURL: function(resource, kind, namespace) {
+      resourceURL: function(resource, kind, namespace, action) {
```
Can we default `action` to `"browse`"?

```
action = action || "browse";
```

jhadvig replied:
+1

spadgett commented on assets/app/views/edit/build-config.html:
```diff
@@ -0,0 +1,471 @@
+                      <div class="form-group" ng-if="strategyType === 'Docker'">
+                        <div class="checkbox">
+                          <label>
+                            <input type="checkbox" ng-model="options.noCache" aria-describedby="from_source_help"/>
```
Remove `aria-describedby`.

jhadvig commented:
@spadgett I've addressed the comments. ptal so we can proceed with the form_ami

spadgett commented on assets/app/scripts/controllers/edit/buildConfig.js:
```diff
@@ -0,0 +1,598 @@
+      switch (type) {
+        case "builder":
+          $scope.builderTagInput = false;
+          var tags = $scope.buildFrom.tags[$scope.options.pickedBuildFromImageStream]
```
;

spadgett commented on assets/app/scripts/controllers/edit/buildConfig.js:
```diff
@@ -0,0 +1,598 @@
+          break;
+        case "output":
+          $scope.outputTagInput = false;
+          var tags = $scope.pushTo.tags[$scope.options.pickedPushToImageStream]
```
;

spadgett commented on assets/app/scripts/controllers/edit/buildConfig.js:
```diff
@@ -0,0 +1,598 @@
+          break;
+        case "imageSource":
+          $scope.imageSourceTagInput = false;
+          var tags = $scope.imageSourceBuildFrom.tags[$scope.options.pickedImageSourceImageStream]
```
;

spadgett commented on assets/app/views/directives/osc-key-values.html:
```diff
@@ -16,7 +16,7 @@
                  type="text"
                  name="value"
-                 placeholder="Value"
+                 placeholder="{{valueTitle}}"
```
Can you try `ng-attr-placeholder` to prevent `{{valueTitle}}` from ever flickering in the field briefly?

spadgett commented:
@jhadvig looks like there's a bindata mismatch

jhadvig commented:
@spadgett already taken care of it and the build of fork_ami is running

jhadvig commented:
@spadgett since the 1.1.1 could be proceed with testing of the pr ?

spadgett commented:
[test]

jhadvig commented:
re-[test] due to flaky tests

jhadvig commented:
@spadgett @jwforres  I've updated the PR, changes:
- imageSource and  buildFrom pickers consist only from selectboxes
- pushTo picker has now input field for the Tag, with `text-warning` when pushing to existing Tag
![screenshot-3](https://cloud.githubusercontent.com/assets/1668218/12558825/fdc1a9d2-c391-11e5-8043-8dfb61101337.png)
- imageChangeTrigger will be disabled if no builderImage is picked in DockerStrategy
![screenshot-6](https://cloud.githubusercontent.com/assets/1668218/12559036/0fb4d7c6-c393-11e5-9fe0-a3b574cb1ac4.png)
- adding `text-danger` (wasn't sure if it shall be `warning` || `danger`) when:
  - picked namespace doesnt contain any IS
![screenshot](https://cloud.githubusercontent.com/assets/1668218/12558772/bd359784-c391-11e5-92e8-3aa81f0d4b58.png)
  - if defined IS is not present in namespaces ISs
![screenshot-2](https://cloud.githubusercontent.com/assets/1668218/12558778/cb11f618-c391-11e5-8cc5-b3938d77131e.png)
  - if picked IS doesn't contain any Tags
![screenshot-1](https://cloud.githubusercontent.com/assets/1668218/12558782/cece9b44-c391-11e5-93cb-45093063d336.png)
- adding `AsFile` input field for binary build
![screenshot-5](https://cloud.githubusercontent.com/assets/1668218/12559046/195c0df8-c393-11e5-886b-93faab8a228c.png)
- fixing: 
  - https://bugzilla.redhat.com/show_bug.cgi?id=1297362
  - https://bugzilla.redhat.com/show_bug.cgi?id=1300644
  - https://bugzilla.redhat.com/show_bug.cgi?id=1301440


PTAL

jwforres commented:
maybe... "Tag 'latest' already exists. Other builds may be pushing to this tag."

jwforres commented:
Does it make sense to include namespaces that can't be selected because there are no image streams in that namespace?  I think I would exclude those namespaces from the dropdown.  Same for image streams with no tags (for the from, not the to).

jhadvig commented:
I thought about it, but what if the user specifies would have in his BC specified builder, but wouldnt create the builder IS. Then he would be wondering why he can see his current namespace in the `Namespace` selectbox. This was he will be noticed whats wrong. The same for the Tags.

jwforres commented on assets/app/views/edit/build-config.html:
```diff
@@ -0,0 +1,475 @@
+                            Binary Input As File
+                            <span class="help action-inline">
+                              <a href>
+                                <i class="pficon pficon-help" data-toggle="tooltip" data-placement="right" aria-hidden="true" data-original-title="Indicates that the provided binary input should be considered a single file within the build input. For example, specifying 'webapp.war' would place the provided binary as '/webapp.war' for the builder. If left empty, the Docker and Source build strate ...
```
can you make the part specific to the custom strategy only show up if it is a custom strategy build?   are you doing validation on this field? since we know the filename can't contain slashes or be ".."

jwforres commented:
the binary input as file is kind of a big scary empty field right at the top of the form... i know you are only going to hit it if you've already chosen to do a binary build, so you are already a bit of an advanced user... but maybe this particular field needs a help-block instead of a tooltip? or it needs to have placeholder text? @spadgett other ideas?

spadgett commented:
> Does it make sense to include namespaces that can't be selected because there are no image streams in that namespace? I think I would exclude those namespaces from the dropdown. Same for image streams with no tags (for the from, not the to).

I think we're lazy loading the image streams -- @jhadvig correct? -- so we'd have to prefetch everything.

spadgett commented:
What happens when I edit a build config that references an image stream that doesn't exist?

spadgett commented:
> wasn't sure if it shall be warning || danger

Typically, if it prevents you from submitting the form, it's an error (`text-danger`). Otherwise, it's a warning.

jhadvig commented:
@spadgett it wont let you to save the form until you fill alt the fields that are requested, except the Tag for the Output Image. You will get just errors that no IS are imported/IS is not present in namespace/...

spadgett commented:
> the binary input as file is kind of a big scary empty field right at the top of the form... i know you are only going to hit it if you've already chosen to do a binary build, so you are already a bit of an advanced user... but maybe this particular field needs a help-block instead of a tooltip? or it needs to have placeholder text? @spadgett other ideas?

Maybe a short help-block description with a "Learn more" link to the docs?

https://docs.openshift.org/latest/dev_guide/builds.html#binary-source

Should we be showing a Dockerfile editor for binary source builds? Or a field for `contextDir`?

We should probably add any fields to browse build config view that we show in the editor.

spadgett commented:
> @spadgett it wont let you to save the form until you fill alt the fields that are requested, except the Tag for the Output Image. You will get just errors that no IS are imported/IS is not present in namespace/...

What happens if the image stream exists, but you don't have permission to get the image streams in the namespace?

liggitt commented:
Yup, need to handle 404 and 403 errors

jhadvig commented:
> Should we be showing a Dockerfile editor for binary source builds? Or a field for contextDir?
Only binary and git source are mutually exclusive. All other combinations of sources should be valid

jwforres commented:
Mmm true... ok.

On Monday, January 25, 2016, Sam Padgett <notifications@github.com> wrote:

> Does it make sense to include namespaces that can't be selected because
> there are no image streams in that namespace? I think I would exclude those
> namespaces from the dropdown. Same for image streams with no tags (for the
> from, not the to).
>
> I think we're lazy loading the image streams -- @jhadvig
> <https://github.com/jhadvig> correct? -- so we'd have to prefetch
> everything.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/6528#issuecomment-174626640>.
>

jhadvig commented:
> What happens if the image stream exists, but you don't have permission to get the image streams in the namespace?

Is it even possible to have permissions only to a subset of ISs in a namespace?

liggitt commented:
Anything is possible. You could have a custom role that does not allow listing image streams, but does allow you to get/pull/push to a particular one.

jhadvig commented:
@liggitt so basically when LISTing of ISs returns 404 we should attempt to GET the standalone IS ?

liggitt commented:
I'm missing context... is this to populate suggestions into a drop-down, or validate a specific user-entered value before you submit it? If the former, not being allowed to list (403) means you don't have a way to scan for the values. If the latter, I would probably try the GET, warn if it returns a 404 or 403, but allow them to submit it anyway.

spadgett commented:
It's to populate suggestions in the dropdown. If the build config points to a value you can't see, you won't be able to save it.

liggitt commented:
If they can't list them, they should be able to enter a namespace/name manually. I would also warn but allow them to submit ones that use values we can't validate... use cases are:
- image streams that are not set up yet, but will be (at which point the build will work)
- image streams that have been exposed to service accounts in the user's project, but not the user themselves (build will work because the service account can push/pull)

jhadvig commented:
@spadgett dont we then need a way to specify that IS, IST ? I mean not just selectBoxes. Referring to the solution we already had with selectbox switching to input field.
Or the YAML editor should be efficient ? I can't help myself but I liked the previous solution much more then this one.

jhadvig commented:
> We should probably add any fields to browse build config view that we show in the editor.

@spadgett would that mean to also add the Dockerfile itself ?

benjaminapetersen commented on assets/app/scripts/controllers/create/createFromImage.js:
```diff
@@ -12,13 +12,14 @@ angular.module("openshiftConsole")
     var displayNameFilter = $filter('displayName');
     var humanize = $filter('humanize');
 
     $scope.projectName = $routeParams.project;
```
Good call getting this out into a constant!

benjaminapetersen commented on assets/app/scripts/controllers/edit/buildConfig.js:
```diff
@@ -0,0 +1,561 @@
+            $scope.strategyType = $scope.buildConfig.spec.strategy.type;
+            $scope.envVars = $filter('envVarsPair')($scope.buildStrategy.env);
+
+            angular.forEach($scope.buildConfig.spec.triggers, function(value, key) {
```
Wonder if this logic could be pushed up into a service.  Perhaps something like:

```javascript
$scope.triggers = BuildConfigService.getTriggers(buildConfig);
```

jhadvig replied:
:+1: on both

benjaminapetersen commented on assets/app/scripts/controllers/edit/buildConfig.js:
```diff
@@ -0,0 +1,561 @@
+              }
+            });
+            $scope.pickedTriggers = $scope.triggers;
+
```
Same here, perhaps in a service?
```javascript
$scope.sources = BuildConfigService.getSources(buildConfig);
```

benjaminapetersen commented on assets/app/scripts/controllers/edit/buildConfig.js:
```diff
@@ -0,0 +1,561 @@
+                  break
+              }
+            });
+
```
Could some of this be pushed up into a service as well?
```javascript
$scope.options.pickedBuild = BuildConfigService.buildFromVars(buildConfig);  //  perhaps return an object like {from: {type:'',namespace:'',imageStream:'',dockerImage:""}}
$scope.options.pickedPush = BuildConfigService.pushToVars(buildConfig); 
```
I oversimplified in my sketch code, but it could be nice to reduce some logic in controller here.

benjaminapetersen commented on assets/app/scripts/controllers/edit/buildConfig.js:
```diff
@@ -0,0 +1,561 @@
+      $scope.options.pickedPushToImageStreamTag = _.find(tags, function(tag) { return tag == "latest" }) || tags[0] || "latest";
+    }
+
+    $scope.save = function() {
```
I tend to like to see save functions for controllers delegate to services as well.  Can still return the `DataService.update()` promise to give control to UI:

```javascript
$scope.save = function() {
 // BuildConfigService update will return the underlying DataService.update('buildconfigs', buildConfigName, ...)
 BuildConfigService
    .update(buildConfigName, $scope.updatedBuildConfig) // hide the details of processing this from the controller
   .then(function() {
     // add success alert
   }, function() {
     // add error alert
   });
}
```

benjaminapetersen commented on assets/app/scripts/controllers/edit/buildConfig.js:
```diff
@@ -0,0 +1,561 @@
+      }
+
+      // Construct updated builder image object based on it's kind
+      var from = {};
```
`var` is function scoped... don't know if it will consistently complain, but prob should declare `from` only once in the function.

benjaminapetersen commented on assets/app/scripts/controllers/edit/buildConfig.js:
```diff
@@ -0,0 +1,561 @@
+      DataService.list("imagestreams", {namespace: projectName}, function(imageStreams) {
+        $scope.pushTo.imageStreams = [];
+        $scope.pushTo.tags = {};
+        var projectImageStreams = imageStreams.by("metadata.name");
```
This chunk of code is duplicated in `$scope.updateImageSourceImageStreams()`, can it be extracted into a separate function? I'm ok with this function calling a few smaller functions.

benjaminapetersen commented on assets/app/scripts/controllers/edit/buildConfig.js:
```diff
@@ -0,0 +1,561 @@
+        var projectImageStreams = imageStreams.by("metadata.name");
+        if (!_.isEmpty(projectImageStreams)) {
+          angular.forEach(projectImageStreams, function(imageStream, name) {
+            var tagList = [];
```
tagList is created in 3 functions, could probably be a function:
```javascript
var tagList = tagList(imageStream);
```

benjaminapetersen commented on assets/app/scripts/controllers/edit/buildConfig.js:
```diff
@@ -0,0 +1,561 @@
+        $scope.updatedBuildConfig.spec.output.to = to;
+      }
+
+      // Update envVars
```
could possibly make `updateEnvVars()` into a function rather than comments?

benjaminapetersen commented on assets/app/scripts/controllers/edit/buildConfig.js:
```diff
@@ -0,0 +1,561 @@
+      });
+      $filter('buildStrategy')($scope.updatedBuildConfig).env = updateEnvVars;
+
+      // Update triggers
```
`updateTriggers()` ?

jhadvig replied:
@benjaminapetersen will decompose and put as much of the logic into a buildConfig service. Thanks Ben !

benjaminapetersen replied:
Awesome!

benjaminapetersen commented:
Nice job, there is a lot of work here.  I sprinkled a few comments around, mostly in the `edit/buildConfig.js` controller.  I like to see us pushing business logic up into services & making our controllers lean.  There is a lot of complexity here, so any decomposition to help declare intent is really helpful.

spadgett commented:
Moving some of the logic into services would help us write unit tests.

@jhadvig I'm fine with a BuildConfigsService, which I know you started to add in #6749.

spadgett commented:
> @spadgett would that mean to also add the Dockerfile itself ?

I think you show be able to view on the browse page anything we let you edit in the web console. Since we have a Dockerfile editor now, we should show the Dockerfile in read-only mode somewhere.

Ace has a read-only mode that I think you can set like this:

```
<div ui-ace readonly></div>
```

csrwng commented on assets/app/scripts/controllers/edit/buildConfig.js:
```diff
@@ -0,0 +1,561 @@
+      {
+        "id": "DockerImage",
+        "title": "Docker Image Link"
+      }
```
we also allow ImageStreamImage in the from: https://github.com/openshift/origin/blob/master/pkg/build/api/validation/validation.go#L327-L365

spadgett commented:
@jhadvig Can you make sure this style is applied to other places we use ace? Is there any class ace editor adds we can use?

https://github.com/openshift/origin/blob/master/assets/app/styles/_core.less#L16-L24

Otherwise, the page zooms in when focusing the editor in iOS.

jhadvig commented:
@spadgett not aware of any class. 
you mean just the `font-size: 16px;` ?

spadgett commented:
Yeah, maybe we can add an `ace-editor` class to all of our editors. If `font-size` isn't `16px` or larger, iPhones when zoom in when the control is focused.

spadgett commented:
It looks like there already is an `.ace_editor` class. We should be able to change core.less to use that instead of `.modal-resource-edit .editor`.

jhadvig commented:
@spadgett ok, will add that as soon as I'm finished with the reworking the picker to selectize.. which turned out to be a bit complicated then I thought :-/

jhadvig commented:
@spadgett I've updated the PR. Also addressed most of the comments. Now when user wants to access an a namespace that is not in the list of the retrieved namespaces, he will get warning.
PTAL

spadgett commented on assets/app/views/edit/build-config.html:
```diff
@@ -0,0 +1,519 @@
+                                required>
+                              </select>
+                              <div class="warnings">
+                                <span class="text-danger" ng-if="!availableProjects.contains(options.pickedImageSourceNamespace)">Selected Namespace is not available.</span>
```
I'd probably just say "Could not load image streams from namespace." and not show the alert. It's odd to see the error twice. Perhaps add a tooltip with error details.

spadgett replied:
Although you might need to fixes in #6353 for the tooltip :(

benjaminapetersen commented on assets/app/scripts/controllers/edit/buildConfig.js:
```diff
@@ -0,0 +1,679 @@
+      });
+    }
+
+    $scope.save = function() {
```
Nice progress on all of this :smile: 

I'm still concerned about the size of methods like `$scope.save` & the number of things happening, particularly for long term maintenance...

spadgett commented on assets/app/views/edit/build-config.html:
```diff
@@ -0,0 +1,519 @@
+                                required>
+                              </div>
+                              <div class="warnings">
+                                <span class="text-warning" ng-if="pushTo.tags[options.pickedPushToImageStream].contains(options.pickedPushToImageStreamTag)">Tag <b>'{{options.pickedPushToImageStreamTag}}'</b> is already created and will be overridden on build.</span>
```
I'd hide this warning if the namespace, image stream, and tag haven't been changed. If you give each field a `name` attribute, you can test `form.myField.$dirty` for each.

You might pull the logic into a scope method if it gets too long.

```
scope.showOutputTagWarning = function() {
 ...
};
```

and

```
<span class="text-warning" ng-if="showOutputTagWarning()">...</span>
```

spadgett commented on assets/app/scripts/controllers/edit/buildConfig.js:
```diff
@@ -0,0 +1,679 @@
+    $scope.emptyMessage = "Loading...";
+    $scope.sourceURLPattern = SOURCE_URL_PATTERN;
+    $scope.options = {};
+    $scope.openshiftBuilderImages = ["ruby","python","perl","php","nodejs","wildfly"]
```
We shouldn't have these hard-coded. You should look instead to see if the image has the builder tag. You can see the logic for that in create.js.

jhadvig replied:
Yeah wanted to go that way, but the problem was that most images(if they are custom) dont have annotation at all

spadgett replied:
They should be in `spec.tags.tags` I thought? We use the builder tag in the add to project catalog.

I'd rather show all than have a magic list. This is a problem if there is a valid builder we're not showing. What happens if I edit a build config that uses an image not in this array? We need to handle that case even if we update to use the builder tag.

spadgett commented on assets/app/styles/_core.less:
```diff
@@ -860,6 +860,47 @@ a.disabled-link {
 }
 
+.dockerfile-editor {
+  min-height: 200px;
```
Is there a reason not to use `height: 200px;`?

spadgett commented on assets/app/styles/_core.less:
```diff
@@ -19,8 +19,8 @@ html, body {
-  .modal-resource-edit .editor {
-    font-size: 16px;
+  .ace_editor {
+    font-size: 16px !important;
```
Is `.console-os .ace_editor` specific enough to avoid the `!important`?

spadgett commented on assets/app/styles/_core.less:
```diff
@@ -860,6 +860,47 @@ a.disabled-link {
+    padding-bottom: 15px;
+  }
+  .warnings {
+    text-align: center;
```
I'd just use Bootstrap class `text-center` if there are no other warning styles applied.

spadgett replied:
Is centering working? It looks like the text is left-aligned (which I'm fine with).

jhadvig replied:
yep its working, will replace it. thanks !

spadgett commented on assets/app/scripts/services/buildConfigs.js:
```diff
@@ -0,0 +1,81 @@
+'use strict';
+
+angular.module("openshiftConsole")
+  .factory("BuildConfigsService", function(){
```
My opinion: These methods all seem really tightly coupled to the controller. If they don't make sense outside the edit build controller, I don't think they should be put in a service.

I know you just pulled these out of the controller based on review feedback.... @benjaminapetersen and I might need to duke it out....

jhadvig replied:
So shall I put them back ?

jhadvig replied:
@spadgett forgot to include the screenshot:
![screenshot-7](https://cloud.githubusercontent.com/assets/1668218/12689535/c725ba2c-c6dd-11e5-980d-73a84e989d69.png)

spadgett replied:
> So shall I put them back ?

I would :-/

The functions only make sense inside the controller, so I would keep them there.

benjaminapetersen replied:
I'm definitely a fan of pushing a lot more of our stuff into services than we do, but I can live with keeping these particular functions in the controller if we will not use them anywhere else, per @spadgett  (Side note, BuildConfigService can just `return { foo: func, bar: func }` and skip the prototype stuff).  

Reducing the large functions down to smaller functions with a single responsibility is really the big thing I linger on.

jhadvig commented:
@spadgett @benjaminapetersen I've updated the PR.
- refactored the whole bc edit controller as much as possible
- imageSource are now handled as array so multiple of them could be edited based on the https://github.com/openshift/origin/pull/6546 change
![screenshot-8](https://cloud.githubusercontent.com/assets/1668218/12751603/4edb4922-c9be-11e5-9926-041170c13ea1.png)
- added dockerfile and imageSource  to the bc browse page
![screenshot-9](https://cloud.githubusercontent.com/assets/1668218/12751632/71390982-c9be-11e5-9492-17db907304a4.png)
![screenshot-10](https://cloud.githubusercontent.com/assets/1668218/12751633/72b28c66-c9be-11e5-8309-d71ffd6fa99a.png)

PTAL

@spadgett fyi I'm still getting this https://github.com/openshift/origin/pull/6528#issuecomment-169763115 error even if I do suggested fix below the comment

spadgett commented:
We might want to remove the left sidebar when editing to discourage navigating away. I'm afraid people won't notice the save button if it's off screen and lose their changes.

liggitt commented:
There are a few places we probably want to protect from accidental navigation (here, during the "Add to project" flows, and when editing yaml)

liggitt commented:
Can be done as a follow-up, but we're at the point where we should add it, I think. Something reusable that can be told the state is dirty and warn before allowing navigation

spadgett commented:
For image sources (browse page), I'd suggest not numbering them. Maybe something like:

Image Sources:

openshift/myimage:latest
Paths: /path/to/dir -> dir

openshift/another:latest
Paths: /foo -> bar

spadgett commented:
Dockerfile editor looks good. I think I'd put a border around it since it's white-on-white and maybe a bit more top margin.

benjaminapetersen commented:
Nice job breaking down the controller!  :smile:

jhadvig commented:
@spadgett I've updated the PR as we have spoken.
Here are the screenshots:

- dockerfile editor with border 
![screenshot-14](https://cloud.githubusercontent.com/assets/1668218/12769214/3693406a-ca16-11e5-9577-71f5e9f2afea.png)

- BC browse page with the the imageSource info (not sure how I feel about it though :-/ ...
![screenshot-12](https://cloud.githubusercontent.com/assets/1668218/12769231/5e9aca42-ca16-11e5-85e1-245715a90ba3.png)

- edit page for the BC with multiple imageSources
![screenshot-13](https://cloud.githubusercontent.com/assets/1668218/12769239/782aa91e-ca16-11e5-9377-b668af0232d6.png)

Rest I've reverted so for the single imageSource the edit page looks like  https://cloud.githubusercontent.com/assets/1668218/12355178/4a023eaa-bb9b-11e5-82d0-c0a6a3e4cea4.png

PTAL

spadgett commented on assets/app/scripts/controllers/edit/buildConfig.js:
```diff
@@ -0,0 +1,752 @@
+
+      // Construct updated builder image object based on it's kind
+      if ($scope.builderOptions.pickedType === "None") {
+        delete $filter('buildStrategy')($scope.updatedBuildConfig).from
```
Needs a `;` or test assets will fail.

spadgett commented on assets/app/scripts/controllers/edit/buildConfig.js:
```diff
@@ -0,0 +1,752 @@
+      if ($scope.outputOptions.pickedType === "None") {
+        // If user will change the output reference to 'None' shall the potential PushSecret be deleted as well?
+        // This case won't delete them.
+        delete($scope.updatedBuildConfig.spec.output.to)
```
```
delete $scope.updatedBuildConfig.spec.output.to;
```

spadgett commented on assets/app/scripts/controllers/edit/buildConfig.js:
```diff
@@ -0,0 +1,752 @@
+      if ($scope.availableProjects.contains(namespace)) {
+        return true;
+      }
+      return false;
```
```
return $scope.availableProjects.contains(namespace);
```

spadgett commented on assets/app/scripts/controllers/edit/buildConfig.js:
```diff
@@ -0,0 +1,752 @@
+      if (imageStreams.length !== 0 && !imageStreams.contains(pickedImageStream)) {
+        return "noMatch";
+      }
+      return "";
```
Feels like these should be two different methods that return booleans rather than one that returns string.

spadgett commented on assets/app/scripts/controllers/edit/buildConfig.js:
```diff
@@ -0,0 +1,752 @@
+        } 
+        if (tagHash[pickedImageStream].length !== 0 && !tagHash[pickedImageStream].contains(pickedTag)) {
+          return "noMatch";
+        }
```
Same comment as above.

spadgett commented on assets/app/views/browse/build-config.html:
```diff
@@ -32,6 +31,9 @@
                     ><i class="fa fa-play fa-fw" aria-hidden="true"></i> Start Build</a>
                   </li>
                   <li>
+                    <a ng-href="{{buildConfig | editResourceURL}}" role="button"><i class="fa fa-pencil fa-fw" aria-hidden="true"></i>Edit</a>
```
Add a space between the icon and the word Edit so the menu items line up correctly.

spadgett commented on assets/app/views/edit/build-config.html:
```diff
@@ -0,0 +1,526 @@
+                                  Automatically build a new image when the builder image changes
+                                  <span class="help action-inline">
+                                    <a href>
+                                      <i class="pficon pficon-help" data-placement="right" aria-hidden="true"
```
These tooltips should have `data-placement="left"` (or perhaps top or bottom) since they're near the right edge of the screen.

spadgett commented on assets/app/views/edit/build-config.html:
```diff
@@ -0,0 +1,526 @@
+            <alerts alerts="alerts"></alerts>
+            <div ng-if="!loaded">Loading...</div>
+
+            <div class="tile" ng-show="buildConfig">
```
Remove `class="tile"` now that you've rebased to pick up the new styles.

spadgett commented on assets/app/views/edit/build-config.html:
```diff
@@ -0,0 +1,526 @@
+                      <div class="buttons gutter-top-bottom gutter-top-bottom-2x">
+                        <button class="btn btn-primary btn-lg"
+                            ng-click="save()"
+                            ng-disabled="form.$invalid || disableInputs">
```
Let's add `form.$pristine` here so the user has to change something before save is enabled.

```
ng-disabled="form.$invalid || form.$pristine || disableInputs"
```

spadgett commented on assets/app/scripts/controllers/edit/buildConfig.js:
```diff
@@ -0,0 +1,752 @@
+            break;
+          case "contextDir":
+            sourceMap.contextDir = true;
+            break;
```
I think this can just be

```
sourceMap[key] = true
```

spadgett commented on assets/app/views/edit/build-config.html:
```diff
@@ -0,0 +1,526 @@
+                                }" ng-model="updatedBuildConfig.spec.source.dockerfile" class="dockerfile-editor dockerfile-mode"></div>
+                              </div>
+
+                              <div class="form-group" ng-if="!!updatedBuildConfig.spec.strategy.dockerStrategy.dockerfilePath">
```
The `!!` aren't necessary.

```
ng-if="updatedBuildConfig.spec.strategy.dockerStrategy.dockerfilePath"
```

spadgett commented:
bindata conflict

spadgett commented on assets/app/views/edit/build-config.html:
```diff
@@ -0,0 +1,526 @@
+<project-header class="top-header"></project-header>
+  <project-page>
+    <!-- Middle section -->
+    <div class="middle-section" ng-if="buildConfig">
```
This `ng-if` reintroduces the runtime error parsing key/values. It needs to be `ng-show`.

spadgett replied:
We should look at the problem with the key/value directive that causes the error, but that is unrelated to your change and can be a follow on.

jhadvig replied:
Thanks @spadgett :+1:

spadgett commented on assets/app/views/edit/build-config.html:
```diff
@@ -0,0 +1,526 @@
+            <alerts alerts="alerts"></alerts>
+            <div ng-if="!loaded">Loading...</div>
+
+            <div ng-show="buildConfig">
```
This `ng-show` is redundant since you have `ng-if="buildConfig"` above.

spadgett commented on assets/app/views/edit/build-config.html:
```diff
@@ -0,0 +1,526 @@
+
+            <div ng-show="buildConfig">
+              <h1>
+                Edit Build Config {{buildConfig.metadata.name}} - {{strategyType}} Build Strategy
```
Nit: I think the strategy type should be small text

```
Edit Build Config {{buildConfig.metadata.name}} <small>{{strategyType}} Build Strategy</small>
```

spadgett commented on assets/app/scripts/controllers/edit/buildConfig.js:
```diff
@@ -0,0 +1,733 @@
+      ImageStreamTag: "Image Stream Tag",
+      ImageStreamImage: "Image Stream Image",
+      DockerImage: "Docker Image Link"
+    }
```
Missing `;`, test-assets might fail.

spadgett commented:
[merge]

jhadvig commented:
@spadgett :v:
